staging/src/k8s.io/component-base/metrics/testutil/metrics.go

```diff
@@ -176,13 +176,94 @@ type Histogram struct {
 	*dto.Histogram
 }
 
-// GetHistogramFromGatherer collects a metric from a gatherer implementing k8s.io/component-base/metrics.Gatherer interface.
+// HistogramVec wraps a slice of Histogram.
+// Note that each Histogram must have the same number of buckets.
+type HistogramVec []*Histogram
+
+// GetSampleCount aggregates the sample count of each inner Historgram.
+func (vec HistogramVec) GetSampleCount() uint64 {
+	var count uint64
+	for _, hist := range vec {
+		count += hist.GetSampleCount()
+	}
+	return count
+}
+
+// GetSampleSum aggregates the sample sum of each inner Historgram.
+func (vec HistogramVec) GetSampleSum() float64 {
+	var sum float64
+	for _, hist := range vec {
+		sum += hist.GetSampleSum()
+	}
+	return sum
+}
+
+// Quantile first aggregates inner buckets of each Histogram, and then
+// computes q-th quantile of a cumulative histogram.
+func (vec HistogramVec) Quantile(q float64) float64 {
+	var buckets []bucket
+
+	for i, hist := range vec {
+		for _, bckt := range hist.Bucket {
+			if i == 0 {
+				buckets = append(buckets, bucket{
+					count:      float64(bckt.GetCumulativeCount()),
+					upperBound: bckt.GetUpperBound(),
+				})
+			} else {
+				buckets[i].count += float64(bckt.GetCumulativeCount())
+			}
+		}
+	}
+
+	if len(buckets) == 0 || buckets[len(buckets)-1].upperBound != math.Inf(+1) {
+		// The list of buckets in dto.Histogram doesn't include the final +Inf bucket, so we
+		// add it here for the rest of the samples.
+		buckets = append(buckets, bucket{
+			count:      float64(vec.GetSampleCount()),
+			upperBound: math.Inf(+1),
+		})
+	}
+
+	return bucketQuantile(q, buckets)
+}
+
+// Average computes wrapped histograms' average value.
+func (vec HistogramVec) Average() float64 {
+	return vec.GetSampleSum() / float64(vec.GetSampleCount())
+}
+
+// Clear clears all fields of the wrapped histograms.
+func (vec HistogramVec) Clear() {
```

Comment:
should Clear just clear the internal histograms or should it go back to an empty array? For example, let's say there's a HistogramVec with {attempts=1, attempts=2} as the two histograms.  After a test, we want to Clear() it, but if we only clear the internal histograms and don't empty the slice (`*vec = nil` with `func (vec *HistogramVec) Clear()` as the function signature), we'll still have the same two labels. WDYT?

Reply:
Clear() doesn't do what we expect here. I opened #99472 fixing it.

Reply:
@adtac I guess that _may_ be a problem of current logic, so once we switch to #99472 - which calls `legacyregistry.Reset()`:

https://github.com/kubernetes/kubernetes/blob/0e32c43e07dfa9f03f7695ad892bf8235f2e33f1/staging/src/k8s.io/component-base/metrics/registry.go#L305-L312

The Reset() will (IIUC) do the things you mentioned - call the underlying dto.Histogram to reset:

```go
func (m *Histogram) Reset()         { *m = Histogram{} }
```

Let me know if it is what you meant.

Reply:
No, what I meant was: while the Reset() function should reset each histogram, if there were 2 preemption attempts, we'll still be left with two empty histograms within the HistogramVec. If the next test has only one preemption attempt, we'll still have 2 labels (the second histogram will be empty, but still present). I'm proposing that we *clear* the HistogramVec array itself instead of clearing each individual underlying Histogram.

Reply:
> if there were 2 preemption attempts, we'll still be left with two empty histograms within the HistogramVec.

Not quite. Each test/workload's collector gets the HistogramVec built from scratch, and previous histograms won't be carried over.

I confirmed by building a test:

- put Preemption testcase ahead of Basic
- run a test with a regex matching Preemption and Basic: `-bench=BenchmarkPerfScheduling/\(Preemption\|SchedulingBasic\)/500Nodes`
- then I got the following result
    ```
    # Preemption testcase
    [DEBUG] metricName = scheduler_e2e_scheduling_duration_seconds
    - [DEBUG] metric.label = [name:"profile" value:"default-scheduler"  name:"result" value:"scheduled" ]
    - [DEBUG] metric.label = [name:"profile" value:"default-scheduler"  name:"result" value:"unschedulable" ]
    [DEBUG] metricName = scheduler_pod_scheduling_duration_seconds
    - [DEBUG] metric.label = [name:"attempts" value:"2" ] <== metric "attempts=1" is not recorded, which is good
    
    # Basic testcase
    [DEBUG] metricName = scheduler_e2e_scheduling_duration_seconds
    - [DEBUG] metric.label = [name:"profile" value:"default-scheduler"  name:"result" value:"scheduled" ] <== makes sense, as basic test doesn't have unschedulable pods
    [DEBUG] metricName = scheduler_pod_scheduling_duration_seconds
    - [DEBUG] metric.label = [name:"attempts" value:"1" ] <== makes sense, metric "attempts=2" is not carried over
    ```

Reply:
Thank you for confirming this, you can resolve this now.